misc: fix spelling mistakes

hzhou opened the pull request:
## Pull Request Description

Add maint/spellcheck.sh and apply it to the codebase. It uses [codespell](https://github.com/codespell-project/codespell) for spell checking.

Fixes #4982

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/all
test:mpich/ch4/all

hzhou commented:
test:mpich/ch4/argobots
test:mpich/ch4/pmi
test:mpich/ch4/xpmem

raffenet commented on test/mpi/datatype/hvecblklen.c:
```diff
@@ -14,7 +14,7 @@
 int main(int argc, char *argv[])
 {
-    MPI_Datatype ot, ot2, newtype;
+    MPI_Datatype dt, dt2, newtype;
```
I don't know the original motivation for using `ot`, but I'm not sure it qualifies as a spelling mistake.

raffenet commented on test/mpi/datatype/vecblklen.c:
```diff
@@ -14,7 +14,7 @@
 int main(int argc, char *argv[])
 {
-    MPI_Datatype ot, ot2, newtype;
+    MPI_Datatype dt, dt2, newtype;
```
same comment about `ot` not being a spelling issue.

raffenet commented on test/mpi/f08/coll/alltoallvf08.f90:
```diff
@@ -8,7 +8,7 @@
       use mpi_f08
       integer ierr, errs
-      integer i, ans, size, rank, color
+      integer i, and, size, rank, color
```
this is not right.

raffenet commented on test/mpi/f08/coll/alltoallwf08.f90:
```diff
@@ -8,7 +8,7 @@
       use mpi_f08
       integer ierr, errs
-      integer i, intsize, ans, size, rank, color
+      integer i, intsize, and, size, rank, color
```
this is not right.

raffenet commented on test/mpi/f08/coll/exscanf08.f90:
```diff
@@ -25,7 +25,7 @@ subroutine uop( cin, cout, count, datatype )
       use mpi_f08
       integer inbuf(2), outbuf(2)
-      integer ans, rank, size
+      integer and, rank, size
```
this is not right.

raffenet commented on test/mpi/f08/coll/nonblockingf08.f90:
```diff
@@ -19,7 +19,7 @@ program main
       TYPE(MPI_Request) req
       integer ierr, errs
-      integer ii, ans
+      integer ii, and
```
not right.

raffenet commented on test/mpi/f08/coll/split_typef08.f90:
```diff
@@ -8,7 +8,7 @@
       use mpi_f08
       integer ierr, errs
-      integer i, ans, size, rank, color
+      integer i, and, size, rank, color
```
not right.

raffenet commented on test/mpi/f08/datatype/typesubf08.f90:
```diff
@@ -13,7 +13,7 @@ program main
       integer fullarr(maxn,maxm),subarr(maxn-3,maxm-4)
       integer i,j, ssize
-      integer size, rank, ans
+      integer size, rank, and
```
not right.

raffenet commented on test/mpi/f08/io/ioharness.defn:
```diff
@@ -46,11 +46,11 @@ enddo
 do i=1, n
-    ans = r*n + (k-1)*n*s + i-1
-    if (buf(i) .ne. ans) then
+    and = r*n + (k-1)*n*s + i-1
```
not right.

raffenet commented on test/mpi/f08/rma/winaccf08.f90:
```diff
@@ -14,7 +14,7 @@ program main
       parameter (nrows=25,ncols=10)
       integer buf(1:nrows,0:ncols+1)
-      integer ans
+      integer and
```
not right.

raffenet commented on test/mpi/f08/rma/winfencef08.f90:
```diff
@@ -14,7 +14,7 @@ program main
       parameter (nrows=25,ncols=10)
       integer buf(1:nrows,0:ncols+1)
-      integer ans
+      integer and
```
not right.

raffenet commented on test/mpi/f08/rma/wingetf08.f90:
```diff
@@ -14,7 +14,7 @@ program main
       parameter (nrows=25,ncols=10)
       integer buf(1:nrows,0:ncols+1)
-      integer ans
+      integer and
```
not right.

raffenet commented on test/mpi/f08/rma/winscale1f08.f90:
```diff
@@ -14,7 +14,7 @@ program main
       parameter (nrows=25,ncols=10)
       integer buf(1:nrows,0:ncols+1)
-      integer ans
+      integer and
```
not right.

raffenet commented on test/mpi/f08/rma/winscale2f08.f90:
```diff
@@ -14,7 +14,7 @@ program main
       parameter (nrows=25,ncols=10)
       integer buf(1:nrows,0:ncols+1)
-      integer ans
+      integer and
```
not right.

raffenet commented on test/mpi/f77/io/ioharness.defn:
```diff
@@ -46,11 +46,11 @@ enddo
 do i=1, n
-    ans = r*n + (k-1)*n*s + i-1
-    if (buf(i) .ne. ans) then
+    and = r*n + (k-1)*n*s + i-1
```
not right.

raffenet commented on test/mpi/f90/f90types/createf90types.c:
```diff
@@ -51,38 +51,38 @@ static int checkType(const char str[], int p, int r, int f90kind, int err, MPI_D
             printf("Wrong combiner type (got %d, should be %d) for %s\n", combiner, f90kind, str);
         } else {
-            int parms[2];
+            int params[2];
```
this doesn't seem right.

hzhou replied:
I'll rename it to `ints` since we use `ints` in everywhere else.

raffenet commented on test/mpi/threads/comm/idup_nb.c:
```diff
@@ -47,7 +47,7 @@ MTEST_THREAD_RETURN_TYPE test_intracomm(void *arg)
     int root, bcastbuf;
     int rank, size;
-    int ans[4], expected[4];
+    int and[4], expected[4];
```
not right.

raffenet commented on test/mpi/threads/pt2pt/ibsend.c:
```diff
@@ -10,7 +10,7 @@
    NUMSENDS from NUMTHREADS sender threads, that
    use [i]bsend/[i]send to send a message of size MSGSIZE
-   to its right neigbour or rank 0 if (my_rank==comm_size-1), i.e.
+   to its right neighbour or rank 0 if (my_rank==comm_size-1), i.e.
```
neighbor

raffenet reviewed (changes requested):
If this will need re-review, it needs to be broken up into smaller commits. Github made me click on individual files at some point because the diff was way too big. This took well over and hour to get through.

raffenet commented on src/include/mpir_topo.h:
```diff
@@ -73,7 +73,7 @@ int MPIR_Topo_canon_nhb_count(MPIR_Comm * comm_ptr, int *indegree, int *outdegre
 int MPIR_Topo_canon_nhb(MPIR_Comm * comm_ptr,
                         int indegree, int sources[], int inweights[],
-                        int outdegree, int dests[], int outweights[]);
+                        int outdegree, int dests[], int outweighs[]);
```
outweights is correct.

hzhou commented:
> 
> 
> If this will need re-review, it needs to be broken up into smaller commits. Github made me click on individual files at some point because the diff was way too big. This took well over an hour to get through.

Thanks @raffenet . I have updated some of the skip rules in `maint/spellcheck.sh`, and also pre-applied some manual spelling fix (e.g. British to US English). Now I think it addresses all the places you pointed out and the script is clean now. As you complained, I split the changes into 3 commits and on my browser seems GitHub is able to load each commit in its entirety.

Hopefully, this is a one-time thing and we will be able to keep the code spelling-clean :) 

Could you review it again?

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/custom
netmod: ch4:ofi
compiler: intel

raffenet commented on src/mpi/romio/adio/ad_lustre/ad_lustre_wrcoll.c:
```diff
@@ -1106,7 +1106,7 @@ static void ADIOI_LUSTRE_IterateOneSided(ADIO_File fd, const void *buf, int *str
      * (stripeSize*number of aggs).  Iteratively call ADIOI_OneSidedWriteAggregation
      * for each segment to aggregate the data to the collective buffers, but only do
-     * the actual write (via flushCB stripe parm) once stripesPerAgg stripes
+     * the actual write (via flushCB stripe param) once stripesPerAgg stripes
```
This is an odd one. The struct is named `stripeParms`, which appears to be short for stripe parameters. So I think either param is technically correct, but parm is better in context.

hzhou replied:
I have whitelisted `parm`, I don't know how this one leaked through. Anyway, I thought `parm` is just an abbreviation of `param` which is an abbreviation of `parameter`. Is `parm` more preferred or does it have a different meaning?

Meanwhile, let me clean up that instance.

raffenet replied:
Personally I prefer "param", but the code in this instance historically uses "parm". I suggest we match the code instead of changing variable names in this instance.

hzhou replied:
Oh, I *manually* did the quoted fix before I decided to whitelist `parm`. So the script is working after all. I have reverted the wrong *manual* change.

raffenet reviewed (commented):
Everything but my one comment looks good. Thanks for splitting up. This was way faster to review this time.

hzhou commented:
@raffenet Rebased and (I believe) it should be good now. Could you take one more look?

hzhou commented:
test:mpich/custom
netmod: ch4:ofi
compiler: intel
